x/fee/keeper/keeper.go

```diff
@@ -0,0 +1,201 @@
+package keeper
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
+	ibcfeetypes "github.com/cosmos/ibc-go/v4/modules/apps/29-fee/types"
+	channeltypes "github.com/cosmos/ibc-go/v4/modules/core/04-channel/types"
+	ibckeeper "github.com/cosmos/ibc-go/v4/modules/core/keeper"
+	"github.com/tendermint/tendermint/libs/log"
+
+	"github.com/neutron-org/neutron/x/fee/types"
+)
+
+type (
+	Keeper struct {
+		cdc        codec.BinaryCodec
+		bankKeeper types.BankKeeper
+		storeKey   storetypes.StoreKey
+		memKey     storetypes.StoreKey
+		paramstore paramtypes.Subspace
+		ibcKeeper  *ibckeeper.Keeper
+	}
+)
+
+func NewKeeper(
+	cdc codec.BinaryCodec,
+	storeKey,
+	memKey storetypes.StoreKey,
+	ps paramtypes.Subspace,
+	ibcKeeper *ibckeeper.Keeper,
+	bankKeeper types.BankKeeper,
+) *Keeper {
+	// set KeyTable if it has not already been set
+	if !ps.HasKeyTable() {
+		ps = ps.WithKeyTable(types.ParamKeyTable())
+	}
+
+	return &Keeper{
+
+		cdc:        cdc,
+		storeKey:   storeKey,
+		memKey:     memKey,
+		paramstore: ps,
+		ibcKeeper:  ibcKeeper,
+		bankKeeper: bankKeeper,
+	}
+}
+
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+func (k Keeper) LockFees(ctx sdk.Context, payer sdk.AccAddress, packetID channeltypes.PacketId, fee ibcfeetypes.Fee) error {
+	k.Logger(ctx).Debug("Trying to lock fees", "packetID", packetID, "fee", fee)
+	store := ctx.KVStore(k.storeKey)
+
+	if _, ok := k.ibcKeeper.ChannelKeeper.GetChannel(ctx, packetID.PortId, packetID.ChannelId); !ok {
+		return sdkerrors.Wrapf(channeltypes.ErrChannelNotFound, "channel with id %s and port %s not found", packetID.ChannelId, packetID.PortId)
+	}
+
+	if err := k.checkFees(ctx, fee); err != nil {
+		return sdkerrors.Wrapf(err, "failed to lock fees")
+	}
+
+	feeInfo := types.FeeInfo{
+		Payer: payer.String(),
+		Fee:   fee,
+	}
+	bzFeeInfo := k.cdc.MustMarshal(&feeInfo)
+	store.Set(types.GetFeePacketKey(packetID.ChannelId, packetID.PortId, packetID.Sequence), bzFeeInfo)
+
+	return k.bankKeeper.SendCoinsFromAccountToModule(ctx, payer, types.ModuleName, fee.Total())
+}
+
+func (k Keeper) DistributeAcknowledgementFee(ctx sdk.Context, receiver sdk.AccAddress, packetID channeltypes.PacketId) {
+	k.Logger(ctx).Debug("Trying to distribute ack fee", "packetID", packetID)
+	feeInfo, err := k.GetFeeInfo(ctx, packetID)
+	if err != nil {
+		k.Logger(ctx).Error("no fee info", "error", err)
+		return
+	}
+
+	cacheCtx, writeFn := ctx.CacheContext()
+
+	// try to distribute ack fee
```

Comment:
it seems like this error should be absolutely impossible. if we have an error here, something terrible happened; maybe it's best to panic here?